python/cudf_polars/cudf_polars/experimental/distinct.py

```diff
@@ -0,0 +1,149 @@
+# SPDX-FileCopyrightText: Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Multi-partition Distinct logic."""
+
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING
+
+import pylibcudf as plc
+
+from cudf_polars.dsl.expressions.base import Col, NamedExpr
+from cudf_polars.dsl.ir import Distinct
+from cudf_polars.experimental.base import PartitionInfo
+from cudf_polars.experimental.dispatch import lower_ir_node
+from cudf_polars.experimental.utils import _lower_ir_fallback
+
+if TYPE_CHECKING:
+    from collections.abc import MutableMapping
+
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.dispatch import LowerIRTransformer
+    from cudf_polars.utils.config import ConfigOptions
+
+
+def lower_distinct(
+    ir: Distinct,
+    child: IR,
+    partition_info: MutableMapping[IR, PartitionInfo],
+    config_options: ConfigOptions,
+) -> tuple[IR, MutableMapping[IR, PartitionInfo]]:
+    """
+    Lower a Distinct IR into partition-wise stages.
+
+    Parameters
+    ----------
+    ir
+        The Distinct IR node to lower.
+    child
+        The reconstructed child of ``ir``. May differ
+        from ``ir.children[0]``.
+    partition_info
+        A mapping from all unique IR nodes to the
+        associated partitioning information.
+    config_options
+        GPUEngine configuration options.
+
+    Returns
+    -------
+    new_node
+        The lowered node (usually Distinct).
```

Comment:
It can wrap around reparation or shuffle nodes but the root is always distinct, no?

Reply:
Yeah, I was thinking that fallback may be different, but I suppose the root will still be Distinct.